src/sources/aws_kinesis_firehose.rs

```diff
@@ -0,0 +1,714 @@
+                            })
+                            .map_err(warp::reject::custom)
+                    })
+                },
+            )
+    }
+
+    fn emit_received() -> impl Filter<Extract = (), Error = warp::reject::Rejection> + Clone {
+        warp::any()
+            .and(warp::header::optional("X-Amz-Firehose-Request-Id"))
+            .and(warp::header::optional("X-Amz-Firehose-Source-Arn"))
+            .map(|request_id: Option<String>, source_arn: Option<String>| {
+                emit!(AwsKinesisFirehoseRequestReceived {
+                    request_id: request_id.as_deref(),
+                    source_arn: source_arn.as_deref(),
+                });
+            })
+            .untuple_one()
+    }
+
+    /// If there is a configured access key, validate that the request key matches it
+    fn authenticate(
+        configured_access_key: Option<String>,
+    ) -> impl Filter<Extract = (), Error = warp::Rejection> + Clone {
+        warp::any()
+            .and(warp::header("X-Amz-Firehose-Request-Id"))
+            .and(warp::header::optional("X-Amz-Firehose-Access-Key"))
+            .and_then(move |request_id: String, access_key: Option<String>| {
+                let configured_access_key = configured_access_key.clone();
+                async move {
+                    match (access_key, configured_access_key) {
+                        (_, None) => Ok(()),
+                        (Some(configured_access_key), Some(access_key))
+                            if configured_access_key == access_key =>
+                        {
+                            Ok(())
+                        }
+                        (Some(_), Some(_)) => {
+                            Err(warp::reject::custom(RequestError::AccessKeyInvalid {
+                                request_id,
+                            }))
+                        }
+                        (None, Some(_)) => {
+                            Err(warp::reject::custom(RequestError::AccessKeyMissing {
+                                request_id,
+                            }))
+                        }
+                    }
+                }
+            })
+            .untuple_one()
+    }
+
+    /// Maps RequestError and warp errors to AWS Kinesis Firehose response structure
+    async fn handle_firehose_rejection(
+        err: warp::Rejection,
+    ) -> Result<impl warp::Reply, Infallible> {
+        let request_id: Option<String>;
+        let message: String;
+        let code: StatusCode;
+
+        if let Some(e) = err.find::<RequestError>() {
+            message = format!("{}", e);
+            code = e.status();
+            request_id = e.request_id();
+        } else if let Some(e) = err.find::<warp::reject::MissingHeader>() {
+            code = StatusCode::BAD_REQUEST;
+            message = format!("Required header missing: {}", e.name());
+            request_id = None;
+        } else {
```

Comment:
I'm not terribly happy about catching the other errors here and marshaling as their debug format. The intention is to transform those errors into the expected JSON response structure, but there doesn't appear to be a way to generically handle them: https://github.com/seanmonstar/warp/issues/451. Instead, I tried to just catch the common errors and let the rest fall through to here.

Theoretically AWS wouldn't be sending requests that would fail for any reason other than the access key not matching, but I could see users hitting the header errors if they are just testing out the source via 'curl' or the like.